### PR TITLE
Added multiple executables support for running and diffing tests in GC Infra

### PR DESCRIPTION
**Pull request by ivdiazsa**
# Summary

This PR adds the feature to run multiple test executables (in the form of dll's) and compare results between them. This is done through a new field in the bench _yaml_ files containing the path(s) to this(these) executable(s).

From the user side, the way of running the infra had no changes. The `suite-create` command automatically looks for the _GCPerfSim_ built during the infra's setup steps, and adds it to the bench _yaml_ files. The only action the user needs to take is to set the `vary` field to `executable` if they so desire to do that diff, as `coreclr` remains the default value for that field.

It is important to mention that rebuilding GCPerfSim will always overwrite the previous one. To test your changes and ensure everything still works properly, make sure you back up the original GCPerfSim build (the one you built when you first setup GC infra), before rebuilding. Keep note of the paths because you will need them for the _yaml_ bench files.

# Bench Files

Let's suppose for this example, that you made some changes to GCPerfSim and want to make sure nothing broke in the tests. For simplicity of this example, we will only be using one Core_Root since the main purpose is to compare the GCPerfSim executables. We will label the unchanged GCPerfSim as `orig_gcperfsim` and the modified one as `mod_gcperfsim`. The bench _yaml_ files now would look something like this:

```yaml
vary: executable
test_executables:
  orig_gcperfsim: <path to the GCPerfSim dll without the changes>
  mod_gcperfsim: <path to the GCPerfSim dll with the changes>
coreclrs:
  a:
    core_root: <path to the CoreCLR core root>
options:
   <option configuration such as timeouts, number of iterations, etc>
common_config:
   <configuration values such as number of heaps, concurrent gcs, etc>
benchmarks:
  2gb:
    arguments:
      tc: 10
      tagb: 300
      tlgb: 2
      <other parameters for GCPerfSim>
  <other benchmark tests to run>
```

# Error Reporting During Runs

The error reporting mechanism was also updated. Now, whenever a test fails, it will also show which executable was running aside of the configuration, coreclr, etc. It now looks like shown here:

```text
*WARNING*: One or more tests in the suite encountered errors.

*** Here is a summary of the problems found: ***

========= <Yaml Filename> =========

======= <Executable Name> =======

===== <Coreclr Name> =====

=== <Configuration Name> ===

- <Benchmark Name> -
<Iteration Number>
<Error Message>

<Stack Trace>

<Repeat with other errors that might have occurred>
```

# Comparing Results

When one runs the `diff` command, the `executable` value can now be set to the `vary` field in the _yaml_ file as shown above. A shortened output of how this comparison looks like is shown here:

```text
Diff of base = orig_gcperfsim and new = mod_gcperfsim

                          ┌────────────────────────────┐
                          │ Summary of important stats │
                          └────────────────────────────┘


                                       │    PctTimePausedInGC │ FirstToLastGCSeconds
                                  name │ Base │  New │ % Diff │ Base │  New │ % Diff
  ─────────────────────────────────────┼──────┼──────┼────────┼──────┼──────┼───────
  DESKTOP-S0L8UMK__a__only_config__2gb │ 33.0 │ 30.7 │  -6.96 │ 31.6 │ 30.2 │  -4.46



                                       │ HeapSizeBeforeMB_Mean │ HeapSizeAfterMB_Mean
                                  name │ Base │  New │  % Diff │ Base │  New │ % Diff
  ─────────────────────────────────────┼──────┼──────┼─────────┼──────┼──────┼───────
  DESKTOP-S0L8UMK__a__only_config__2gb │ 5282 │ 5277 │ -0.0881 │ 4070 │ 4060 │ -0.233



                                       │ PauseDurationMSec_95PWhereIsGen0 │ PauseDurationMSec_95PWhereIsGen1
                                  name │ Base │  New │             % Diff │ Base │  New │             % Diff
  ─────────────────────────────────────┼──────┼──────┼────────────────────┼──────┼──────┼───────────────────
  DESKTOP-S0L8UMK__a__only_config__2gb │ 45.3 │ 41.6 │              -8.23 │  100 │ 95.9 │              -4.54



                                       │ PauseDurationMSec_95PWhereIsBackground │ PauseDurationMSec_95PWhereIsBlockingGen2
                                  name │ Base │  New │                   % Diff │        Base │         New │       % Diff
  ─────────────────────────────────────┼──────┼──────┼──────────────────────────┼─────────────┼─────────────┼─────────────
  DESKTOP-S0L8UMK__a__only_config__2gb │ 40.2 │ 36.4 │                    -9.62 │ <no values> │ <no values> │



                                 ┌──────────────────────────────────────┐
                                 │ DESKTOP-S0L8UMK__a__only_config__2gb │
                                 └──────────────────────────────────────┘


Improvements (Improvement of 5-20%)

                                          Metric │      Base (run 0) │       New (run 0) │  % Diff │ Abs Diff
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  CountIsBackground                              │                13 │                12 │   -7.69 │       -1
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  FirstEventToFirstGCSeconds                     │             0.655 │             0.615 │   -6.08 │  -0.0398
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_95P                          │              91.2 │              84.7 │   -7.20 │    -6.57
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_95PWhereIsBackground         │              40.2 │              36.4 │   -9.62 │    -3.87
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_95PWhereIsGen0               │              45.3 │              41.6 │   -8.23 │    -3.73
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_Mean                         │              36.9 │              32.9 │   -10.8 │    -3.98
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_MeanWhereIsEphemeral         │              37.3 │              33.2 │   -11.2 │    -4.17
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationSeconds_Sum                       │              19.5 │              17.4 │   -10.8 │    -2.10
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationSeconds_SumWhereIsBackground      │              9.45 │              8.48 │   -10.3 │   -0.973
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationSeconds_SumWhereIsGen1            │              5.49 │              4.99 │   -8.99 │   -0.493
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationSeconds_SumWhereIsNonBackground   │              10.0 │              8.89 │   -11.3 │    -1.13
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PctTimeInGC_WhereIsNonBackground               │              31.7 │              29.4 │   -7.14 │    -2.26
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PctTimePausedInGC                              │              33.0 │              30.7 │   -6.96 │    -2.30
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  speed                                          │              53.7 │              50.5 │   -5.84 │    -3.13


Stale (Same, or percent difference within 5% margin)

                                          Metric │      Base (run 0) │       New (run 0) │  % Diff │ Abs Diff
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  CountIsBlockingGen2                            │                 0 │                 0 │       0 │        0
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  CountIsGen0                                    │               189 │               189 │       0 │        0
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  CountIsGen1                                    │                81 │                81 │       0 │        0
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  CountUsesLOHCompaction                         │ <not implemented> │ <not implemented> │         │
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  FirstToLastGCSeconds                           │              31.6 │              30.2 │   -4.46 │    -1.41
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  HeapCount                                      │                 6 │                 6 │       0 │        0
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  HeapSizeAfterMB_Max                            │              4419 │              4498 │    1.79 │     79.3
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  HeapSizeAfterMB_Mean                           │              4070 │              4060 │  -0.233 │    -9.47
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  HeapSizeBeforeMB_Max                           │              5673 │              5749 │    1.33 │     75.6
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  HeapSizeBeforeMB_Mean                          │              5282 │              5277 │ -0.0881 │    -4.65
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_95PWhereIsBlockingGen2       │       <no values> │       <no values> │         │
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationMSec_95PWhereIsGen1               │               100 │              95.9 │   -4.54 │    -4.56
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationSeconds_SumWhereIsBlockingGen2    │                 0 │                 0 │       0 │        0
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PauseDurationSeconds_SumWhereUsesLOHCompaction │ <not implemented> │ <not implemented> │         │
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PctIsEphemeral                                 │              95.4 │              95.7 │   0.355 │    0.338
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PctReductionInHeapSize_Mean                    │              22.8 │              22.9 │   0.515 │    0.117
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PctUsesCompaction                              │              93.6 │              94.0 │   0.355 │    0.332
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PromotedMB_MeanWhereIsBlockingGen2             │       <no values> │       <no values> │         │
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PromotedMB_MeanWhereIsGen0                     │              51.7 │              51.8 │   0.114 │   0.0588
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  PromotedMB_MeanWhereIsGen1                     │               229 │               228 │  -0.420 │   -0.963
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  TotalAllocatedMB                               │          3.23e+05 │          3.22e+05 │  -0.217 │     -701
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  TotalLOHAllocatedMB                            │              8.01 │              8.01 │       0 │        0
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  TotalNonGCSeconds                              │              22.2 │              21.9 │   -1.51 │   -0.335
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  TotalNumberGCs                                 │               283 │               282 │  -0.353 │       -1
  ───────────────────────────────────────────────┼───────────────────┼───────────────────┼─────────┼─────────
  TotalSecondsTaken                              │              32.7 │              31.2 │   -4.58 │    -1.49

<numbers from other configurations, coreclrs, and benchmarks run>
```

When one makes changes to GCPerfSim, we want these differences between original and modified be as small as possible to ensure the new changes did not break existing functionality or made it somehow malfunction.

# Other Changes

All the necessary documentation was updated with the addition of this new parameter, and also some broken links were fixed.

### CODE REVIEW

**Comment by Maoni0** [comment]
is the yaml you gave meant as an example for this feature? 'cause it looks like it's still varying coreclr, not GCPerfSim.

**Comment by ivdiazsa** [comment]
> is the yaml you gave meant as an example for this feature? 'cause it looks like it's still varying coreclr, not GCPerfSim.

@Maoni0 If you mean here in the PR description, then it was a mistake. I've changed it now to `executable`. If you mean in the actual docs, I left it as `coreclr` because that's the default infra creates when `suite-create` is called. `Vary` can have multiple values (one at a time of course), and `coreclr` is the default. I didn't see a need to change that, since before this change, you already had to go and change it manually when for example you wanted to now compare configurations.

**Comment by Maoni0** [comment]
I meant in the PR description 'cause someone will look at the PR to see how to use this. looks like you'll need more update in your example 'cause it's still not using different builds of GCPerfsim.

**Comment by ivdiazsa** [comment]
I'm having a hard time understanding what you mean here. Are you referring to `docs/example.md`? That example is to show how to perform basic runs and basic analysis without messing with the tool. I can add another one showing comparisons between GCPerfSim builds if that's what you're referring to.

**Comment by Maoni0** [comment]
no, I was referring to to the PR description - the one you changed to 

vary: executable
test_executables:
  defgcperfsim: <path to the GCPerfSim dll built in the earlier steps>
coreclrs:
  a:
    core_root: <path to first CoreCLR core root>
  b:
    core_root: <path to second CoreCLR core root>

this is still not demonstrating how you would add 2 different GCPerfSims to test

**Comment by ivdiazsa** [comment]
@Maoni0 I've updated the documentation and the PR description to better explain this feature :)

File: src/benchmarks/gc/README.md [hunk 1]
```diff
@@ -226,9 +233,9 @@ Running this produced a directory called `bench/suite/low_memory_container.yaml.
 This contains a trace file (and some other small files) for each of the tests. (If you had specified `collect: none` in `options:` in the benchfile, there would be no trace file and the other files would contain all information.)
 Each trace file can be opened in PerfView if you need to.
 
-Each trace file will be named `{coreclr_name}__{config_name}__{benchmark_name}__{iteration}`, e.g.  `clr_a__smaller__nosurvive__0`.
+Each trace file will be named `{executable_name}__{coreclr_name}__{config_name}__{benchmark_name}__{iteration}`, e.g.  `defgcperfsim__clr_a__smaller__nosurvive__0`.
```
Comment (Maoni0) [inline src/benchmarks/gc/README.md:236]: is defgcperfsim intentional or is it supposed to be just gcperfsim?

Comment (Maoni0) [reply src/benchmarks/gc/README.md:236]: ahh never mind, I see your defgcperfsim example below

**Review by Maoni0 — APPROVED** [review]
LGTM!

**Pull request merged**